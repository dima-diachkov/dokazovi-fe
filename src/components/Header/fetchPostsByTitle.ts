import axios from 'axios';
import { IPost } from '../../old/lib/types';

interface IPostsResponse {
  content: IPost[];
  empty: boolean;
  first: boolean;
  last: boolean;
  number: number;
  numberOfElements: number;
  pageable: any;
  size: 100;
  sort: any;
  totalElements: number;
  totalPages: number;
}

const resultsAmount = 100;
const baseUrl = `https://dokazovi-be-release.herokuapp.com:443/api/post/all-posts?size=${resultsAmount}&title=`;

export const fetchPostsByTitle = async (title: string): Promise<IPost[]> => {
  const postsData = await axios.get<IPostsResponse>(`${baseUrl}${title}`);
  const posts: IPost[] = postsData.data.content;

  return posts;
};
