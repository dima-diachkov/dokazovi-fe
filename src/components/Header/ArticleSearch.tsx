import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import { useSelector } from 'react-redux';
import TextField from '@material-ui/core/TextField';
import Popper from '@material-ui/core/Popper';
import Box from '@material-ui/core/Box';
import { useActions } from 'shared/hooks';
import InputAdornment from '@material-ui/core/InputAdornment';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';
import { Autocomplete } from '@material-ui/lab';
import SearchIcon from '@material-ui/icons/Search';
import { useStyles } from './Header.styles';
import { useEffectExceptOnMount } from '../../old/lib/hooks/useEffectExceptOnMount';
import { fetchPostsByTitle } from './fetchPostsByTitle';
import { searchTitle } from '../../models/materials/reducers';
import { selectMaterials } from '../../models/materials';

// import { IPostsOBJ } from '../../models/adminLab/types';
// import { IPost } from '../../old/lib/types';

interface IArticleSearch {
  setVisibility: React.Dispatch<React.SetStateAction<boolean>>;
}

interface IOption {
  id: number;
  title: string;
}

export const ArticleSearch: React.FC<IArticleSearch> = ({ setVisibility }) => {
  const classes = useStyles();
  const history = useHistory();
  const [boundedSearchTitle] = useActions([searchTitle]);
  const [postOptions, setPostOptions] = useState<IOption[]>([]);
  const [title, setTitle] = useState('');

  const handleInputChange = (
    event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>,
  ) => {
    setTitle(event.target.value);
  };

  useEffectExceptOnMount(() => {
    if (!title) {
      setPostOptions([]);
      return;
    }
    const getTitles = async () => {
      const posts = await fetchPostsByTitle(title);
      const titleOptions = posts.map((post) => {
        const { id, title: titleOption } = post;
        return { id, title: titleOption };
      });
      setPostOptions(titleOptions);
    };
    getTitles();
  }, [title]);

  const handlePickOption = (e, value) => {
    history.push(`/posts/${value.id}`);
    setVisibility(false);
    boundedSearchTitle({ title: '' });
  };

  const handleSearch = () => {
    if (!title) return;
    setVisibility(false);
    boundedSearchTitle({ title });
    history.push(`/materials/search`);
  };

  const handleClickAway = () => {
    boundedSearchTitle({ title: '' });
    setVisibility(false);
  };

  const PopperMy = (props) => {
    return (
      <Popper
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...props}
        style={{
          marginTop: '22px',
          width: '100%',
          textOverflow: 'ellipsis',
          overflow: 'hidden',
          whiteSpace: 'nowrap',
        }}
      />
    );
  };

  return (
    <ClickAwayListener onClickAway={handleClickAway}>
      <Box className={classes.searchInputWrapper}>
        <Autocomplete
          filterOptions={(x) => x}
          fullWidth
          disableClearable
          freeSolo
          options={postOptions}
          PopperComponent={PopperMy}
          getOptionLabel={(option: IOption) => option.title}
          onChange={(e, value) => handlePickOption(e, value)}
          renderInput={(params) => (
            <TextField
              // eslint-disable-next-line react/jsx-props-no-spreading
              {...params}
              value={title}
              variant="standard"
              className={classes.searchInput}
              onChange={(event) => handleInputChange(event)}
              InputProps={{
                ...params.InputProps,
                endAdornment: (
                  <InputAdornment
                    position="end"
                    className={classes.inputAdornment}
                  >
                    <SearchIcon
                      className={classes.searchInputIcon}
                      onClick={handleSearch}
                    />
                  </InputAdornment>
                ),
                disableUnderline: true,
              }}
            />
          )}
        />
      </Box>
    </ClickAwayListener>
  );
};
