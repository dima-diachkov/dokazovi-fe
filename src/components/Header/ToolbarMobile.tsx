import React, { useState } from 'react';
import Button from '@material-ui/core/Button';
import SearchIcon from '@material-ui/icons/Search';
import { useStyles } from './Header.styles';
import { Logo } from './Logo';
import { BurgerMenu } from '../BurgerMenu/BurgerMenu';
import { navElements } from './navElements';
import { ArticleSearch } from './ArticleSearch';

export const ToolbarMobile: React.FC = () => {
  const classes = useStyles();
  const [isSearchVisible, setIsSearchVisible] = useState<boolean>(false);
  const [mobileMenuOpen, setMobileMenuOpen] = useState<boolean>(false);

  return (
    <>
      <BurgerMenu
        navigation={navElements}
        mobileMenuOpen={mobileMenuOpen}
        setMobileMenuOpen={(b) => setMobileMenuOpen(b)}
      />
      <Logo isOnMobile isSearchVisible={isSearchVisible} />
      {isSearchVisible ? (
        <ArticleSearch setVisibility={setIsSearchVisible} />
      ) : (
        <Button
          aria-controls="simple-menu"
          aria-haspopup="true"
          onClick={() => setIsSearchVisible((prev) => !prev)}
        >
          <SearchIcon className={classes.searchIcon} />
        </Button>
      )}
    </>
  );
};
