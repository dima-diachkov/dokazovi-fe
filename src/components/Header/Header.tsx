import React, { useContext, useState } from 'react';
import { Container, Slide } from '@material-ui/core';
import { useStyles } from './Header.styles';
import { ScreenContext } from '../../old/provider/MobileProvider/ScreenContext';
import { ToolbarMobile } from './ToolbarMobile';
import { ToolbarDesktop } from './ToolbarDesktop';
import { useActions } from '../../shared/hooks';
import {
  makeHeaderInvisible,
  makeHeaderVisible,
} from '../../models/headerVisibility';

export const Header: React.FC = () => {
  const classes = useStyles();
  const [visible, setVisible] = useState<boolean>(true);
  const [oldPageOffsetY, setOldPageOffsetY] = useState<number>(0);
  const { mobile } = useContext(ScreenContext);
  const [boundMakeHeaderVisible, boundMakeHeaderInvisible] = useActions([
    makeHeaderVisible,
    makeHeaderInvisible,
  ]);

  const [searchInputValue, setSearchInputValue] = useState<string>('');

  if (mobile) {
    window.onscroll = () => {
      setOldPageOffsetY(visualViewport.pageTop);

      if (visualViewport.pageTop >= oldPageOffsetY && visible) {
        setVisible(false);
        boundMakeHeaderInvisible();
      }
      if (visualViewport.pageTop <= oldPageOffsetY && !visible) {
        setVisible(true);
        boundMakeHeaderVisible();
      }
    };
  }

  return (
    <Slide in={!mobile || visible} timeout={{ enter: 200, exit: 200 }}>
      <div
        id="header"
        className={mobile ? classes.headerMobile : classes.header}
      >
        <Container className={classes.container}>
          {mobile ? (
            <ToolbarMobile setInput={setSearchInputValue} />
          ) : (
            <ToolbarDesktop />
          )}
        </Container>
      </div>
    </Slide>
  );
};
