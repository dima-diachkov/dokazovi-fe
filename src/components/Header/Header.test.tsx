import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { Header } from './Header';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { AuthContext } from '../../old/provider/AuthProvider/AuthContext';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore({
  currentUser: {
    data: {
      id: 12,
      firstName: 'Sash',
      lastName: 'Nabokin',
      avatar: '',
      qualification: '',
      phone: '123',
      email: '1@ssd.com',
      bio: '',
      mainInstitution: undefined,
      mainDirection: undefined,
      directions: [],
      lastAddedPost: {
        id: 2,
        title: 'Asar',
      },
      socialNetwork: undefined,
    },
  },
});
const history = createMemoryHistory();

test('component renders properly', () => {
  render(
    <Router history={history}>
      <Header />
    </Router>,
  );

  expect(screen.getByTestId('header')).toBeInTheDocument();
  expect(screen.getByTestId('header')).toMatchSnapshot();
});

test('should show login button if user unauthenticated', () => {
  const newTestContext = {
    authenticated: false,
    setAuthorization: () => {},
    removeAuthorization: () => {},
    checkPermission: () => false,
  };
  render(
    <Router history={history}>
      <AuthContext.Provider value={newTestContext}>
        <Header />
      </AuthContext.Provider>
    </Router>,
  );

  expect(screen.getByText('Увійти')).toBeInTheDocument();
});

test('it should show account menu & post creation menu if user authenticated', () => {
  const testContext = {
    authenticated: true,
    setAuthorization: () => {},
    removeAuthorization: () => {},
    checkPermission: () => false,
  };
  render(
    <Provider store={store}>
      <Router history={history}>
        <AuthContext.Provider value={testContext}>
          <Header />
        </AuthContext.Provider>
      </Router>
    </Provider>,
  );

  expect(screen.getByText('Створити...')).toBeInTheDocument();
  expect(screen.getByText('Sash')).toBeInTheDocument();
});
