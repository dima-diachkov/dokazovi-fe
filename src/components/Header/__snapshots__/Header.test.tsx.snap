// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders properly 1`] = `
<div
  class="Header-header-1"
  data-testid="header"
  id="header"
>
  <div
    class="MuiContainer-root Header-container-2 MuiContainer-maxWidthLg"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular Header-toolbar-3 MuiToolbar-gutters"
    >
      <div
        class="MuiBox-root MuiBox-root-9"
      >
        <a
          href="/"
        >
          <h1
            class="MuiTypography-root Header-logo-4 MuiTypography-h1"
          >
            Доказові
          </h1>
        </a>
        <div
          class="MuiBox-root MuiBox-root-10 Header-tabs-6"
        >
          <a
            aria-current="page"
            class="Header-tab-7 active"
            href="/"
          >
            <h5
              class="MuiTypography-root Header-tabLabel-8 MuiTypography-h5"
            >
              Головна
            </h5>
          </a>
          <a
            class="Header-tab-7"
            href="/materials"
          >
            <h5
              class="MuiTypography-root Header-tabLabel-8 MuiTypography-h5"
            >
              Матеріали
            </h5>
          </a>
          <a
            class="Header-tab-7"
            href="/experts"
          >
            <h5
              class="MuiTypography-root Header-tabLabel-8 MuiTypography-h5"
            >
              Автори
            </h5>
          </a>
        </div>
      </div>
      <div
        class="MuiBox-root MuiBox-root-11 Header-actionsContainer-5"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-12 MuiButton-textPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <svg
              class="makeStyles-icon-13"
              viewBox="0 0 55 55"
            >
              <path
                d="M27.7113 53.3466C41.8694 53.3466 53.3469 41.8692 53.3469 27.711C53.3469 13.5529 41.8694 2.07544 27.7113 2.07544C13.5531 2.07544 2.07568 13.5529 2.07568 27.711C2.07568 41.8692 13.5531 53.3466 27.7113 53.3466Z"
                stroke="white"
                stroke-width="1.5"
              />
              <path
                d="M27.7121 31.4115C33.2847 31.4115 37.8021 26.8941 37.8021 21.3215C37.8021 15.7489 33.2847 11.2314 27.7121 11.2314C22.1395 11.2314 17.6221 15.7489 17.6221 21.3215C17.6221 26.8941 22.1395 31.4115 27.7121 31.4115Z"
                stroke="white"
                stroke-width="1.5"
              />
              <path
                d="M45.9864 45.6722C41.8932 40.0025 35.2359 36.302 27.7041 36.302C20.1782 36.302 13.515 39.9965 9.42188 45.6722"
                stroke="white"
                stroke-width="1.5"
              />
            </svg>
            <h5
              class="MuiTypography-root makeStyles-label-14 MuiTypography-h5"
            >
              Увійти
            </h5>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
